### PR TITLE
Increment port by random amount when trying to find a good port

### LINKED ISSUE #40: Test::PostgreSQL is racy when choosing port numbers
When running tests that use Test::PostgreSQL with yath and a high amount
of jobs (reproduced with 16), sometimes tests can fail with colliding
port numbers:

    ( STDERR )  job 11    Failed to start postgres on port 15442: Failed to open /tmp/nix-shell.1ytszg/yath-14161-MqDHWh/tmp/NclbJE/pgtest.prkeW/data/postmaster.pid: No such file or directory at /nix/store/ka5mvpfzgc9z709sx9bljlm7691n2hgm-hydra-perl-deps/lib/perl5/site_perl/5.32.0/Test/PostgreSQL.pm line 429.
    ( STDERR )  job 12    Failed to start postgres on port 15442: Failed to open /tmp/nix-shell.1ytszg/yath-14161-MqDHWh/tmp/xv6Jwh/pgtest.prkeW/data/postmaster.pid: No such file or directory at /nix/store/ka5mvpfzgc9z709sx9bljlm7691n2hgm-hydra-perl-deps/lib/perl5/site_perl/5.32.0/Test/PostgreSQL.pm line 429.

This can be somewhat worked around by setting the base_port with
a random number, but this only decreases the likelihood and doesn't
totally protect against this.

### PR DESCRIPTION
**Pull request by TJC**
Increment port by random amount when trying to find a good port

Attempts to fix #40

### CODE REVIEW

File: lib/Test/PostgreSQL.pm [hunk 1]
```diff
@@ -399,12 +399,14 @@ method _find_port_and_launch() {
...
-        die "Failed to start postgres on port $port: $_";
+        die "Failed to start postgres after trying 10 potential ports: $_";
       }
       undef;
     };
     return if $good;
-    $port++;
+    # Increment port by a random number to avoid clashes with other Test::Postgresql processes
+    # Keep in mind that this increment is going to be made up to 10 times, so avoid exceeding 64k
+    $port += int(rand(500)) + 1;
```
Comment (cole-h) [inline lib/Test/PostgreSQL.pm:409]: I'm not a Perlmonger by any means but: shouldn't we seed with `srand()` first?

**Comment by cole-h** [comment]
Unfortunately, I'm still seeing the issue:

```
( STDERR )  job 13    Failed to start postgres after trying 10 potential ports: Failed to open /tmp/nix-shell.vJktRU/yath-32538-5cteuh/tmp/S_SOTW/pgtest.prkeW/data/postmaster.pid: No such file or directory at /nix/store/7w22g36mp6an5jrrr9ar2k338bsqzyq6-hydra-perl-deps/lib/perl5/site_perl/5.32.0/Test/PostgreSQL.pm line 431.
( STDERR )  job 12    Failed to start postgres after trying 10 potential ports: Failed to open /tmp/nix-shell.vJktRU/yath-32538-5cteuh/tmp/jqZuRy/pgtest.prkeW/data/postmaster.pid: No such file or directory at /nix/store/7w22g36mp6an5jrrr9ar2k338bsqzyq6-hydra-perl-deps/lib/perl5/site_perl/5.32.0/Test/PostgreSQL.pm line 431.
```

File: lib/Test/PostgreSQL.pm (hunk 1 above)
Comment (TJC) [reply lib/Test/PostgreSQL.pm:409]: I haven't written any Perl for several years, so you're probably more of a Perlmonger than I am, at this point.

However I believe you don't need to call srand(), at least not on any modern build of Perl.

**Comment by TJC** [comment]
Is it possible there is another problem on your system, that prevents more than ~10 Postgresqls to exist simultaneously?
(eg. Global shared memory pool sizes)

**Comment by cole-h** [comment]
Not impossible, but extremely unlikely, considering I have 32G of RAM available. I'll do some playing around with various things in that area of the code... Thanks for being so quick to respond!

File: lib/Test/PostgreSQL.pm (hunk 1 above)
Comment (cole-h) [reply lib/Test/PostgreSQL.pm:409]: So, I did a bit more testing and...

By calling `srand()` just before the `while` loop begins, I have yet to see colliding ports. After commenting out the `srand()` call, I immediately see a colliding port.

I don't know why this works... but it works, and it's a simple fix:

```diff
diff --git a/lib/Test/PostgreSQL.pm b/lib/Test/PostgreSQL.pm
index 55728f5..eb75938 100644
--- a/lib/Test/PostgreSQL.pm
+++ b/lib/Test/PostgreSQL.pm
@@ -390,6 +390,7 @@ method start() {
 method _find_port_and_launch() {
   my $tries = 10;
   my $port = $self->base_port;
+  srand();
   # try by incrementing port number until PostgreSQL starts
   while (1) {
     my $good = try {
```

Comment (TJC) [reply lib/Test/PostgreSQL.pm:409]: That's interesting, and confusing.
I'm happy to include that if it's a fix. I'm also curious to understand how you're using the library.

Is your code loading Test::PostgreSQL, and then using fork() to split into many processes?
I think that's a known scenario where Perl's RNG needs reseeding.

**Comment by TJC** [comment]
I've pushed an update including a call to `srand` where you suggested

**Comment by cole-h** [comment]
That works wonders, thank you!

You can find the test suite I'm working with here: https://github.com/NixOS/hydra/tree/de2f30f035636ca5f261a429606fc28d375c0cbc/t, but I can't really say more (I only recently started working on it; Test::PostgreSQL usage here: https://github.com/NixOS/hydra/blob/de2f30f035636ca5f261a429606fc28d375c0cbc/t/lib/Setup.pm#L61).

The tests are run with `yath` (and ports collide with e.g. `yath -j16`) -- maybe that has something to do with it?

**Comment by TJC** [comment]
I have a suspicion that the problem was:
- The code only attempts 10 starts before giving up, and you were running 16 tests at once
- The Perl RNG doesn't re-seed after a fork, so both forks will generate the same random numbers

I'll cut a new release now, after merging this branch.

**Comment by cole-h** [comment]
Ah, right, that makes sense. Thanks a ton!

**Pull request merged**